packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/sharedobjects/SharedObjectRegistry.kt

```diff
@@ -43,15 +44,20 @@ class SharedObjectRegistry(appContext: AppContext) {
     // It's already defined in the JS base SharedObject class prototype,
     // but with the current implementation it's possible to use a raw object for registration.
     js.defineProperty(sharedObjectIdPropertyName, id.value)
-    appContextHolder
-      .get()
-      ?.jsiInterop
-      ?.setNativeStateForSharedObject(id.value, js)
+
+    val appContext = appContextHolder.get() ?: throw Exceptions.AppContextLost()
+
+    appContext
+      .jsiInterop
+      .setNativeStateForSharedObject(id.value, js)
 
     val jsWeakObject = js.createWeak()
     synchronized(this) {
       pairs[id] = native to jsWeakObject
     }
+
+    native.appContextHolder = WeakReference(appContext)
```

Comment:
The registry already has `appContextHolder`, can we reuse it here?
```suggestion
    native.appContextHolder = appContextHolder
```

Reply:
I didn't want to reuse the same ref from the registry because someone could invoke `clear,` which would remove the references in multiple places - that might be hard to debug.